src/wallet/wallet.cpp

```diff
@@ -137,6 +135,8 @@ void UnloadWallet(std::shared_ptr<CWallet>&& wallet)
     // Notify the unload intent so that all remaining shared pointers are
     // released.
     wallet->NotifyUnload();
+    wallet->BlockUntilSyncedToCurrentChain();
+    wallet->m_chain_notifications_handler.reset();
```

Comment:
This is a partial fix for the bug, not a full fix. While calling BlockUntilSyncedToCurrentChain before reset makes it less likely the wallet pointer will be deleted while still in use, it's still possible for a blockconnected or other notification to be running when wallet.reset() is called below

Reply:
That's guaranteed it https://github.com/bitcoin/bitcoin/pull/18354/commits/16e9578c1ecc3240282e136c378fb0433e775e41#diff-e8d9e22d9683f73a9fb8399be0dab640R93

Reply:
As I understand it, the EmptyQueue call added there (in UnregisterValidationInterface) does nothing if `m_pscheduler->AreThreadsServicingQueue()` is true, which it always is, except during shutdown. So a notification could still be in progress and raw wallet pointer still in use after the UnregisterValidationInterface call and handler.reset() call and wallet.reset() call. Adding the BlockUntilSyncedToCurrentChain right before probably makes this much less unlikely, but not impossible. #18338 is be a more direct, race-free fix for the issue, because it gets rid of the raw pointer and uses shared_ptr for all wallet references.